### PR TITLE
tests: Improving 'make check' execution time

### PR DESCRIPTION
**Pull request by ErwanAliasr1**
When running make -j x check, we face a weird situation where the makefile
targets are spawn in parallel up to "x" but some targets is very very
long and sequential.

This pull request aims at removing this bottlenecks to make a faster build process.

The initial running time was 30mn on a recent laptop and dropped to 14mn.

Signed-off-by: Erwan Velu <erwan@redhat.com>

### CODE REVIEW

File: src/test/encoding/readable.sh [hunk 1]
```diff
@@ -114,15 +118,86 @@ for arversion in `ls -v $dir/archive`; do
...
+     for pid in "$@"; do
+       shift
+       if kill -0 "$pid" 2>/dev/null; then
+         debug "$pid is still alive."
+         set -- "$@" "$pid"
+       elif wait "$pid"; then
+         debug "$pid exited with zero exit status."
+       else
+         debug "$pid exited with non-zero exit status."
+         ((++errors))
```
Comment (tchaikov) [inline src/test/encoding/readable.sh]: this is bash extension not POSIX compatible. in other words, this might not work in some distro, for instance debian derivatives, which point `/bin/sh` to `dash`.

**Comment by ErwanAliasr1** [comment]
@tchaikov Repushed with the fixes you asked for. I rebased & updating the commit comment to explain that we can override the number of // jobs by using MAX_PARALLEL_JOBS

**Comment by dachary** [comment]
:+1:

**Comment by branch-predictor** [comment]
:+1:, here it reduced make check time from 28 down to 17 minutes.

**Comment by wjwithagen** [comment]
On 15-3-2016 17:15, Erwan Velu wrote:
> When running make -j x check, we face a weird situation where the makefile
> targets are spawn in parallel up to "x" but one of those target is very very
> long and sequential.
>
> The "readable.sh" test is trying to run ~7.9K tests where 5.3K are actually
> executed.
>
> The current code is taking 23mn on a recent laptop (Intel(R) Core(TM)
> i7-4810MQ CPU @ 2.80GHz, 32GB of RAM & SSD).
>
> This patch implements parallelism to speed up this process which is not
> really CPU and
> neither IO bound.
>
> By default, readable.sh is now using the number of logical processors to
> determine
> the level of parallelism (by using nproc).
>
> On the same system, where nproc=8, the resulting execution time is 5m55
> seconds :
> 4x faster than the original code.
>
> The global 'make check' is therefore getting faster too and dropped from
> 30 to
> 16 minutes : 2x faster than the original code.

Something I've contemplating is to separate the checks into at least three
parts:
	simple (unit) tests with no benchmark component
	convoluted tests which require ods and mons to run
	benchmarks that really take a long time.

And then have make run these tests in sequence.
Problem for me now is that benchmarks and complexer tests are run, whilest
there could be trouble in the more simpel tests.

Just haven't gotten around it. Next to that unittests are not always without
benchmark part and/or requiring osds/mons

--WjW

**Comment by ErwanAliasr1** [comment]
Just added a new commit to save one more minute without impacting the loadavg.

**Comment by ErwanAliasr1** [comment]
Please don't merge that PR now, I found many other options to optimize the make check. I'll update the PR accordingly and tell when it's ready.

**Comment by dachary** [comment]
@ErwanAliasr1 we usually prefix the title of a pull request that must not be merged with DNM (Do Not Merge). When it's ready, just remove the prefix :-)

File: src/test/encoding/readable.sh [hunk 2]
```diff
@@ -4,23 +4,25 @@ dir=../ceph-object-corpus
...
 
-tmp1=`mktemp /tmp/typ-XXXXXXXXX`
-tmp2=`mktemp /tmp/typ-XXXXXXXXX`
-
 failed=0
 numtests=0
+pids=""
 
 myversion=`./ceph-dencoder version`
+DEBUG=0
```
Comment (tchaikov) [inline src/test/encoding/readable.sh]: @ErwanAliasr1 could you revise the title of your commit message with something like: "test/encoding/readable.sh: parallelize the tests" ?

so it is more specific?

**Comment by tchaikov** [comment]
tested the readable.sh changes: 52m53.915s => 24m20.108s. will continue with other commits tomorrow.

File: src/test/encoding/readable.sh (hunk 2 above)
Comment (ErwanAliasr1) [inline src/test/encoding/readable.sh]: Good point, fixed & repushed.

File: src/test/ceph_objectstore_tool.py [hunk 3]
```diff
@@ -44,9 +44,9 @@ def wait_for_health():
     print "Wait for health_ok...",
     tries = 0
     while call("./ceph health 2> /dev/null | grep -v 'HEALTH_OK\|HEALTH_WARN' > /dev/null", shell=True) == 0:
-        if ++tries == 30:
+        if ++tries == 150:
```
Comment (tchaikov) [inline src/test/ceph_objectstore_tool.py]: this is actually a bug. in python "+" is an identity operator. so "++tries" evaluates to "try".

Comment (ErwanAliasr1) [inline src/test/ceph_objectstore_tool.py]: Fixed by commit https://github.com/ceph/ceph/pull/8131/commits/820ed6b1b000a15a0fc2a8efc954dfb8185ada50

File: src/test/ceph_objectstore_tool.py [hunk 4]
```diff
@@ -44,7 +44,8 @@ def wait_for_health():
     print "Wait for health_ok...",
     tries = 0
     while call("./ceph health 2> /dev/null | grep -v 'HEALTH_OK\|HEALTH_WARN' > /dev/null", shell=True) == 0:
-        if ++tries == 150:
+        tries = tries + 1
```
Comment (tchaikov) [inline src/test/ceph_objectstore_tool.py]: could be
```py
tries += 1
```

and my first name is "Kefu"

also ideally, you could have an empty line before the "Signed-off-by" line. "git commit -s" can do it nicely.

**Comment by ErwanAliasr1** [comment]
Most of the work is done here. I think its time for review + merge.
If I found something else I'll make a 2nd PR for it.

**Comment by ErwanAliasr1** [comment]
Rebased & pushed.

File: qa/workunits/ceph-helpers.sh [hunk 5]
```diff
@@ -197,26 +197,30 @@ function kill_daemons() {
...
     local signal=${2:-TERM}
     local name_prefix=$3 # optional, osd, mon, osd.1
-    local delays=${4:-0 0 1 1 1 2 3 5 5 5 10 10 20 60 60 60 120}
+    local delays=${4:-0.1 0.2 1 1 1 2 3 5 5 5 10 10 20 60 60 60 120}
 
     local status=0
+    pids=""
     for pidfile in $(find $dir 2>/dev/null | grep $name_prefix'[^/]*\.pid') ; do
-        pid=$(cat $pidfile)
+        (pid=$(cat $pidfile)
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: move that part to a separate function, it's becoming too big

File: qa/workunits/ceph-helpers.sh [hunk 6]
```diff
@@ -197,26 +197,30 @@ function kill_daemons() {
...
-        done
-        if ! $kill_complete ; then
-            status=1
-        fi
+            sleep $try
+        done; exit $exit_code) &
+	pids="$pids $!"
+    done
+    for pid in $pids; do
+	if ! wait $pid; then
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: if $pid terminated before wait gets there, wait will return 127 which will be confused as an error when it really means the daemon is already dead

Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: I like this optimization :-) It needs to be tested in test_kill_daemons (the race mentionned above can be faked by making it so the first daemon in the list blocs and the second one terminates quickly).

Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: and the documentation needs updating

File: qa/workunits/ceph-helpers.sh [hunk 7]
```diff
@@ -1032,23 +1036,25 @@ function test_is_clean() {
...
 
+    while true ; do
         cur_active_clean=$(get_num_active_clean)
+        test $cur_active_clean = $num_pgs && break
         if test $cur_active_clean != $num_active_clean ; then
             timer=0
             num_active_clean=$cur_active_clean
+        elif get_is_making_recovery_progress ; then
+            timer=0
+        elif (( timer >= $TIMEOUT )) ; then
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: timer * 10

File: qa/workunits/ceph-helpers.sh [hunk 8]
```diff
@@ -1032,23 +1036,25 @@ function test_is_clean() {
 #
 function wait_for_clean() {
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: the documentation needs updating

File: src/test/ceph_objectstore_tool.py [hunk 9]
```diff
@@ -1796,16 +1797,12 @@ def main(argv):
         vstart(new=False)
         wait_for_health()
 
-        time.sleep(20)
```
Comment (dachary) [inline src/test/ceph_objectstore_tool.py]: I unfortunately think there is a reason for that sleep but it's unclear what it is.

**Comment by dachary** [comment]
to validate this series of change, a single make check is not enough. Since it modifies timing and timeouts in a number of places, make check should be run in a loop for a day or so with no failure.

File: src/test/encoding/check-generated.sh [hunk 10]
```diff
@@ -24,16 +24,21 @@ for type in `./ceph-dencoder list_types`; do
...
 
-	./ceph-dencoder type $type select_test $n dump_json > $tmp1
-	./ceph-dencoder type $type select_test $n encode decode dump_json > $tmp2
-	./ceph-dencoder type $type select_test $n copy dump_json > $tmp3
-	./ceph-dencoder type $type select_test $n copy_ctor dump_json > $tmp4
+	./ceph-dencoder type $type select_test $n dump_json > $tmp1 & pids="$!"
+	./ceph-dencoder type $type select_test $n encode decode dump_json > $tmp2 & pids="$pids $!"
+	./ceph-dencoder type $type select_test $n copy dump_json > $tmp3 & pids="$pids $!"
+	./ceph-dencoder type $type select_test $n copy_ctor dump_json > $tmp4 & pids="$pids $!"
+	for pid in $pids; do wait $pid; done
```
Comment (dachary) [inline src/test/encoding/check-generated.sh]: watch out for the -127 return that indicates the process finished early

File: src/test/encoding/check-generated.sh [hunk 11]
```diff
@@ -65,10 +70,11 @@ for type in `./ceph-dencoder list_types`; do
...
 
-	if ./ceph-dencoder type $type is_deterministic
-	then
-	    ./ceph-dencoder type $type select_test $n encode export $tmp1
-	    ./ceph-dencoder type $type select_test $n encode decode encode export $tmp2
+	if [ $deterministic -ne 0 ]; then
+	    ./ceph-dencoder type $type select_test $n encode export $tmp1 & pids="$!"
+	    ./ceph-dencoder type $type select_test $n encode decode encode export $tmp2 & pids="$pids $!"
+	    for pid in $pids; do wait $pid; done
+
```
Comment (dachary) [inline src/test/encoding/check-generated.sh]: dito

**Comment by dachary** [comment]
the pattern proc & pids+="$!" ... wait for $pids to complete is used in many places and looks useful. It deserves a function in ceph-helpers.sh with various tests for border cases such as when a pid terminates early. There also is a potentially confusing mix of the output of all the process that are run in parallel. We could either print them as they go and prefix each line with the command. Or redirect each stdout/stderr in a separate file and print them one after the other so the developer does not need to sort which output belongs to which command when trying to figure out the source of the problem.

**Comment by wjwithagen** [comment]
@ErwanAliasr1 @dachary 
I did some hacking in qa/workunits/ceph-helpers.sh to adopt to FreeBSD. if you want look at it they are in #7573, but that deserves a big cleanup.

One of the difficult parts there is that is seems that under FreeBSD OSD/MONs only die after 
2* kill TERM.

But I've also become rather annoyed at the fact that with failures the dumps of the OSD/MONs are in the output. Especially during porting/testing/debugging it is much more useful if the tests in testdir are not removed and can be used for post-mortum debugging. I've added tests on the ENV CEPH_DEBUG, but any other way of achieving this would be welcom. 

I also noticed @ldachary mentions "update manual" but in all attempts to find manuals about the testing, I was not able to find anything. (Perhaps because I don't make the manuals during building. But on the Ceph site I did not find much)

File: src/test/ceph_objectstore_tool.py (hunk 4 above)
Comment (ErwanAliasr1) [inline src/test/ceph_objectstore_tool.py]: Fixed. Sorry for the mispelling.

File: qa/workunits/ceph-helpers.sh (hunk 7 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done.

File: qa/workunits/ceph-helpers.sh (hunk 8 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done.

File: src/test/ceph_objectstore_tool.py (hunk 9 above)
Comment (ErwanAliasr1) [inline src/test/ceph_objectstore_tool.py]: I had the exact same remark. But as long the wait_for_health is complete it should be ok. If not, then we shall make a specific test instead of waiting an amount of time which doesn't guarantee anything.
As far as we cannot determine the meaning of such wait, my proposal here is to remove it and add a real test if someone notices a regression. I didn't had any during my tests.

File: src/test/encoding/check-generated.sh (hunk 10 above)
Comment (ErwanAliasr1) [inline src/test/encoding/check-generated.sh]: ceph-dencoder is lasting long so this is very unlikely but if so, we are in the same behavior right ? I mean the return code of the ceph-dencoder wasn't tested before. Shall we ?

File: src/test/ceph_objectstore_tool.py (hunk 9 above)
Comment (dachary) [inline src/test/ceph_objectstore_tool.py]: > if someone notices a regression

this is going to be a painful process since a) the problem is most probably going to be transient, b) the person observing it is not going to figure out this removal could be the cause. 

the removal of these two sleep() is desirable but they deserve a separate pull request, discussion with david and very careful thinking

File: src/test/encoding/check-generated.sh (hunk 10 above)
Comment (dachary) [inline src/test/encoding/check-generated.sh]: the whole script runs with **set -e** and the expected behavior is that it aborts as soon as something fails

Comment (dachary) [inline src/test/encoding/check-generated.sh]: > ceph-dencoder is lasting long so this is very unlikely 

It will happen if **ceph-dencoder type $type select_test $n dump_json** finishes after **ceph-dencoder type $type select_test $n encode decode dump_json**

**Comment by dachary** [comment]
Hum, actually wait $pid will only return -127 if the pid has never been a child of the current shell. So there is nothing special to do to handle -127.

File: src/test/ceph_objectstore_tool.py (hunk 9 above)
Comment (ErwanAliasr1) [inline src/test/ceph_objectstore_tool.py]: @dzafman Can you enlight us on the need of having those sleeps here ?

File: src/test/encoding/check-generated.sh (hunk 10 above)
Comment (ErwanAliasr1) [inline src/test/encoding/check-generated.sh]: Fair enough. That kills my PR then. I have to think on how to handle that.

Comment (ErwanAliasr1) [inline src/test/encoding/check-generated.sh]: We are safe against the return code 127 as per https://gist.github.com/ErwanAliasr1/1969b33d2c538a65d8ed

We only get 127 if the process never existed which is not the case here.

Comment (ErwanAliasr1) [inline src/test/encoding/check-generated.sh]: So for me the code is safe as wait will assert if one pid return failed and will not get 127 because of races between process.

**Comment by ErwanAliasr1** [comment]
@dachary Updated the branch to implement & use helpers. Remains the logging issue to fix.

File: qa/workunits/ceph-helpers.sh [hunk 12]
```diff
@@ -146,6 +146,27 @@ function test_teardown() {
 #######################################################################
 
 ##
+# Kills a single daemon. This is an helper function fo kill_daemons
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: Could you please document it as a first class citizen ? All functions in this file are helpers and they should all be individually tested and documented. The general idea is that someone interested in a given function does not need to look beyond the part of the file where it's defined, documented and tested. That locality is very valuable for the casual user.

File: qa/workunits/ceph-helpers.sh [hunk 13]
```diff
@@ -1023,32 +1033,36 @@ function test_is_clean() {
 
 ##
 # Wait until the cluster becomes clean or if it does not make progress
-# for $TIMEOUT seconds. The function **is_clean** is used to determine
-# if the cluster is clean. Progress is measured either vian the
-# **get_is_making_recovery_progress** predicate or if the number of
-# clean PGs changes.
+# for $TIMEOUT seconds.
+# Comparing get_num_active_clean & get_num_pgs is used to determine
+# if the cluster is clean. That's almost an inline of is_clean() to
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: the documentation is designed to be read by the user and should not contain implementation details. it should focus on documenting precisely each aspect of function behavior and error handling, pitfall etc. 

the reason why is_clean is inlined could usefuly be added as a comment within the function itself, as a reminder for the implementer

File: qa/workunits/ceph-helpers.sh [hunk 14]
```diff
@@ -1282,6 +1296,53 @@ function test_display_logs() {
 }
 
 #######################################################################
+##
+# run_in_background() function spawn the command line passed in
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/run_in_background() function // (like the other descriptions)

File: qa/workunits/ceph-helpers.sh [hunk 15]
```diff
@@ -1282,6 +1296,53 @@ function test_display_logs() {
...
+# argument in background and save the pid in the variable name passed
+# in 1st argument like :
+#
+# pids1=""
+# run_in_background pids1 bash -c 'sleep 1; exit 1'
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: Add a section with @param & @return as for the other fucntions

File: qa/workunits/ceph-helpers.sh [hunk 16]
```diff
@@ -1282,6 +1296,53 @@ function test_display_logs() {
...
+#
+function run_in_background() {
+    # The first argument is the name of the PID variable
+    # We execute everything passed in argument
+    # And save the running pid in the variable name
+    local pid_variable=$1
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: the above comment is redundant with the documentation and can be removed

File: qa/workunits/ceph-helpers.sh [hunk 17]
```diff
@@ -1282,6 +1296,53 @@ function test_display_logs() {
...
+    shift;
+    "$@" & eval "$pid_variable+=\" $!\""
+}
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: add a test_run_in_background() function that verifies it does the right thing

File: qa/workunits/ceph-helpers.sh [hunk 18]
```diff
@@ -1282,6 +1296,53 @@ function test_display_logs() {
...
+    done
+
+    # We empty the variable reporting that all process ended
+    eval "$1=''"
+
+    return $return_code
+}
+
+
+#######################################################################
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: remove that extra separator

File: qa/workunits/ceph-helpers.sh [hunk 19]
```diff
@@ -146,6 +146,27 @@ function test_teardown() {
...
+	    exit_code=0
+            break
+         fi
+         send_signal=0
+         sleep $try
+    done;
+    return $exit_code
+}
+
+##
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: add a test_kill_deamon function (moving code from test_kill_daemons most probably)

File: qa/workunits/ceph-helpers.sh (hunk 12 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done in the new version of that commit.

File: qa/workunits/ceph-helpers.sh (hunk 13 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Fixed in the new version of that commit.

File: qa/workunits/ceph-helpers.sh (hunk 14 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done in last version of the commit.

File: qa/workunits/ceph-helpers.sh (hunk 15 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done in last version of the commit.

File: qa/workunits/ceph-helpers.sh (hunk 16 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done in last version of the commit.

File: qa/workunits/ceph-helpers.sh (hunk 18 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done in last version.

File: qa/workunits/ceph-helpers.sh (hunk 19 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: kill_daemons doesn't do much except calling kill_daemon several time.
The actual test_kill_daemons test the feature pretty well. Having that test passing implies that kill_daemon is ok. Seems overkill to me to add such test.

File: qa/workunits/ceph-helpers.sh (hunk 17 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Added in the last version of the commit.

File: qa/workunits/ceph-helpers.sh [hunk 20]
```diff
@@ -146,6 +146,50 @@ function test_teardown() {
 #######################################################################
 
 ##
+# Sends a signal to a single daemon.
+# This is an helper function fo kill_daemons
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/an/a/ s/fo/for/

File: qa/workunits/ceph-helpers.sh [hunk 21]
```diff
@@ -146,6 +146,50 @@ function test_teardown() {
...
+#
+# After the daemon is sent **signal**, its actual termination
+# will be verified by sending it signal 0. If the daemon is
+# still alive, kill_daemons will pause for a few seconds and
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/daemons/daemon/

File: qa/workunits/ceph-helpers.sh [hunk 22]
```diff
@@ -146,6 +146,50 @@ function test_teardown() {
...
+# try again. This will repeat for a fixed number of times
+# before kill_daemons returns on failure. The list of
+# sleep intervals can be specified as **delays** and defaults
+# to:
+#
+#  0.1 0.2 1 1 1 2 3 5 5 5 10 10 20 60 60 60 120
+#
+# This sequence is designed to not require a sleep time (0) if the
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/not require a sleep time (0)/only sleep 0.1 second/

File: qa/workunits/ceph-helpers.sh (hunk 19 above)
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: You can create the test function by moving some of the test cases from test_kill_daemons. The important point here is that someone maintaining this function should not need to look anywhere else to fix a bug or add a feature. It should be self contained. Sure, the kill_deamons tests also cover kill_daemon. But when a newcomer works on kill_daemon, it is more work for her/him to figure that out. 

This helper file has been carefully designed and maintained to preserve its maintainability and usability.

File: qa/workunits/ceph-helpers.sh [hunk 23]
```diff
@@ -167,24 +211,6 @@ function test_teardown() {
...
-#  0 1 1 1 2 3 5 5 5 10 10 20 60 60 60 120
-#
-# This sequence is designed to not require a sleep time (0) if the
-# machine is fast enough and the daemon terminates in a fraction of a
-# second. The increasing sleep numbers should give plenty of time for
-# the daemon to die even on the slowest running machine. If a daemon
-# takes more than a few minutes to stop (the sum of all sleep times),
-# there probably is no point in waiting more and a number of things
-# are likely to go wrong anyway: better give up and return on error.
-#
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: the @param delay below is no longer valid

File: qa/workunits/ceph-helpers.sh [hunk 24]
```diff
@@ -197,27 +223,17 @@ function kill_daemons() {
     local dir=$1
     local signal=${2:-TERM}
     local name_prefix=$3 # optional, osd, mon, osd.1
-    local delays=${4:-0 0 1 1 1 2 3 5 5 5 10 10 20 60 60 60 120}
-
+    local delays=$4 #optional timing
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: I think this argument is only used by test_kill_daemons and can be removed now that it moved to kill_daemon.

File: qa/workunits/ceph-helpers.sh [hunk 25]
```diff
@@ -1023,32 +1039,36 @@ function test_is_clean() {
 
 ##
 # Wait until the cluster becomes clean or if it does not make progress
-# for $TIMEOUT seconds. The function **is_clean** is used to determine
-# if the cluster is clean. Progress is measured either vian the
-# **get_is_making_recovery_progress** predicate or if the number of
-# clean PGs changes.
+# for $TIMEOUT seconds.
+# Progress is measured either via the **get_is_making_recovery_progress**
+# predicate or if the number of clean PGs changes.
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/number of clean PGs/number of clean PGs as returned by **get_num_active_clean**/

File: qa/workunits/ceph-helpers.sh [hunk 26]
```diff
@@ -1282,6 +1302,70 @@ function test_display_logs() {
...
+##
+# Spawn a command in background and save the pid in the variable name
+# passed in argument
+#
+# Example:
+#   pids1=""
+#   run_in_background pids1 bash -c 'sleep 1; exit 1'
+#
+# @param pid_variable the variable name (not value) where the pids will be stored
+# @param $@ the command to execute
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: **$@** also includes the first argument, use **...** instead (that's how it's done in other functions as well, C prototype style).

File: qa/workunits/ceph-helpers.sh [hunk 27]
```diff
@@ -1282,6 +1302,70 @@ function test_display_logs() {
...
+    done
+
+    # We empty the variable reporting that all process ended
+    eval "$1=''"
+
+    return $return_code
+}
+
+
+function test_run_in_background() {
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: s/test_run_in_background/test_wait_background/

File: qa/workunits/ceph-helpers.sh [hunk 28]
```diff
@@ -1282,6 +1302,70 @@ function test_display_logs() {
...
+#
+# @param pid_variable the variable name (not value) where the pids will be stored
+# @param $@ the command to execute
+# @return only the pid_variable output should be considered and used with **wait_background**
+#
+function run_in_background() {
+    local pid_variable=$1
+    shift;
+    "$@" & eval "$pid_variable+=\" $!\""
+}
```
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: Here it is:

<pre>
function test_run_in_background() {
   local pids
   run_in_background pids sleep 1
   test "$pids" || return 1
   wait $pids || return 1
}
</pre>

File: src/test/encoding/check-generated.sh [hunk 29]
```diff
@@ -17,23 +19,35 @@ for type in `./ceph-dencoder list_types`; do
...
     for n in `seq 1 1 $num 2>/dev/null`; do
-	if ! ./ceph-dencoder type $type select_test $n encode decode; then
+	safe_type=$type
+	# BitVector<2> needs some escaping to avoid bash issues with <>
+	if [ "$type" = "BitVector<2>" ]; then
+	    safe_type="BitVector\<2\>"
+	fi
+
+	pids=""
+	run_in_background pids bash -c "./ceph-dencoder type $safe_type select_test $n dump_json > $tmp1"
```
Comment (dachary) [inline src/test/encoding/check-generated.sh]: You can 

    run_in_background pids ./ceph-dencoder type $safe_type select_test $n dump_json > $tmp1

there is no need for the bash -c. But it does not really matter ;-)

File: src/test/osd/osd-scrub-repair.sh [hunk 30]
```diff
@@ -75,8 +75,13 @@ function corrupt_and_repair_two() {
     #
     # 1) remove the corresponding file from the OSDs
     #
-    objectstore_tool $dir $first SOMETHING remove || return 1
-    objectstore_tool $dir $second SOMETHING remove || return 1
+    pids=""
+    run_in_background pids objectstore_tool $dir $first SOMETHING remove
```
Comment (dachary) [inline src/test/osd/osd-scrub-repair.sh]: should something fail here, it will be very difficult to sort out what comes from the first process and what comes from the second process. I suggest run_in_background redirects stdout/stderr and print them in order, when the process terminates. Since the pids are collected in order, the output will also be displayed in order, which will help a lot to debug.

File: src/test/encoding/readable.sh [hunk 31]
```diff
@@ -114,15 +122,86 @@ for arversion in `ls -v $dir/archive`; do
...
+            running_jobs=$(($running_jobs - 1))
+        done
+        running_jobs=0
+    fi
   done
 done
 
-rm -f $tmp1 $tmp2
-
 if [ $failed -gt 0 ]; then
```
Comment (dachary) [inline src/test/encoding/readable.sh]: this refactor is extensive. since readable.sh is essential to preserve data integrity, it's worth having more reviewers as well as tests that demonstrate it will keep catching the errors it is supposed to catch. it would be bad (that happened to me more than once) to realize in two months time that a slight mistake turned the test into something that succeeds when it should fail ;-)

Comment (dachary) [inline src/test/encoding/readable.sh]: I propose this is turned into a separate pull request as it is relatively independant from the rest.

File: qa/workunits/ceph-helpers.sh (hunk 20 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: That's an helper as it starts with a vowel sound.
Fixed for the for.

File: qa/workunits/ceph-helpers.sh (hunk 21 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: thx, fixed.

File: qa/workunits/ceph-helpers.sh (hunk 24 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: It was possible to pass this argument to kill_daemons so I was propagating the value. Is that an issue ? That still could be useful.

File: qa/workunits/ceph-helpers.sh (hunk 25 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: fixed

File: qa/workunits/ceph-helpers.sh (hunk 26 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: fixed thx

File: qa/workunits/ceph-helpers.sh (hunk 28 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Why does mine is incorrect ?

File: qa/workunits/ceph-helpers.sh (hunk 27 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: done

File: qa/workunits/ceph-helpers.sh (hunk 28 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Oh you want a test for both. Ok.

Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: added. thx.

File: qa/workunits/ceph-helpers.sh (hunk 19 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: I will look at that tomorrow

File: qa/workunits/ceph-helpers.sh (hunk 22 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: fixed.

File: src/test/encoding/readable.sh (hunk 31 above)
Comment (ErwanAliasr1) [inline src/test/encoding/readable.sh]: The main goal was a performance gain. But would be nice to get more review on it yes.

File: src/test/encoding/check-generated.sh (hunk 29 above)
Comment (ErwanAliasr1) [inline src/test/encoding/check-generated.sh]: I'm not sure because of the redirection. I don't really check how it would be handled.

File: src/test/osd/osd-scrub-repair.sh (hunk 30 above)
Comment (ErwanAliasr1) [inline src/test/osd/osd-scrub-repair.sh]: Yes I still have to work on the interlaced output stream.

File: qa/workunits/ceph-helpers.sh (hunk 24 above)
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: It's not used for anything but testing the function itself and such testing should now be done in kill_daemon.

File: qa/workunits/ceph-helpers.sh (hunk 28 above)
Comment (dachary) [inline qa/workunits/ceph-helpers.sh]: nit: could you group run_in_background and test_run_in_background together and separate them from wait_background and test_wait_in_background with a ####################################################################### line ? So it is like all other functions in the file. Exactly one function, one test and the documentation :-)

File: src/test/encoding/check-generated.sh (hunk 29 above)
Comment (dachary) [inline src/test/encoding/check-generated.sh]: you can leave it as it is, that's ok

File: qa/workunits/ceph-helpers.sh (hunk 24 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: This have been removed.

File: qa/workunits/ceph-helpers.sh (hunk 28 above)
Comment (ErwanAliasr1) [inline qa/workunits/ceph-helpers.sh]: Done.

File: src/test/osd/osd-scrub-repair.sh (hunk 30 above)
Comment (ErwanAliasr1) [inline src/test/osd/osd-scrub-repair.sh]: Fixed in my latest commit by prepending the output with the process id.

**Comment by ErwanAliasr1** [comment]
I have an issue with one test, I'll fix it asap

File: src/test/ceph_objectstore_tool.py (hunk 9 above)
Comment (dzafman) [inline src/test/ceph_objectstore_tool.py]: @ErwanAliasr1  It is fine to remove this sleep and the sleep(15) below too.

Comment (ErwanAliasr1) [inline src/test/ceph_objectstore_tool.py]: Thx @dzafman.

**Comment by ErwanAliasr1** [comment]
On my side the PR is over & solved all the open issues. @ldachary if you want to make a final check on it, that's ok on my side.

**Comment by ErwanAliasr1** [comment]
The branch have been rebased on the current master too.

**Comment by ErwanAliasr1** [comment]
Rebased against current master tree.

**Comment by dachary** [comment]
:+1: @ErwanAliasr1 could you please rebase one last time to get a clean make check ? I assume the failure does not come from your series.

**Comment by ErwanAliasr1** [comment]
@dachary I'm still up to date with master, I can't rebase anymore

**Comment by ErwanAliasr1** [comment]
@dachary Rebased against master to get fix from PR #8429 . That should be ok this way.

**Comment by dachary** [comment]
Let's hope it's not tainted by http://tracker.ceph.com/issues/15375
